File: apps/runtime/src/git/commit.rs
//! Reading HEAD, committing (including amend) and `git init`.

use super::*;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CommitResult {
    /// Short hash of the commit that was just created.
    pub commit: String,
    pub committed: Vec<String>,
    pub summary: String,
}

/// The commit `--amend` would rewrite, as the composer has to present it.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HeadCommit {
    pub oid: String,
    pub subject: String,
    /// The full message, so an amend can start from it instead of silently
    /// dropping the body. Empty when `truncated`.
    pub message: String,
    /// The stored message is larger than an amend may resend; the composer
    /// must refuse rather than rewrite the commit with a shortened message.
    pub truncated: bool,
    /// At least one remote-tracking ref contains this commit, so rewriting it
    /// rewrites history other checkouts may already have fetched.
    pub published: bool,
}

/// `HEAD` as the amend controls describe it, or None on an unborn branch.
pub fn head_commit(workspace_root: &Path, requested: &str) -> AppResult<Option<HeadCommit>> {
    let context = require_repository(workspace_root, requested)?;
    let Ok(oid) = git(
        &context.repository,
        &["rev-parse", "--verify", "--quiet", "HEAD^{commit}"],
    ) else {
        return Ok(None);
    };
    let oid = oid.trim().to_owned();
    if oid.is_empty() {
        return Ok(None);
    }
    let message = git(
        &context.repository,
        &[
            "log",
            "-1",
            "--no-show-signature",
            "--format=%B",
            &oid,
            "--",
        ],
    )?;
    // Trailing newlines are Git's own formatting, not part of the message.
    let message = message.trim_end_matches('\n').to_owned();
    let truncated = message.len() > 10_000;
    let subject = message.lines().next().unwrap_or_default().to_owned();
    let containing = git(
        &context.repository,
        &[
            "for-each-ref",
            "--format=%(refname)",
            "--contains",
            &oid,
            "refs/remotes/",
        ],
    )?;
    Ok(Some(HeadCommit {
        oid,
        subject,
        message: if truncated { String::new() } else { message },
        truncated,
        published: !containing.trim().is_empty(),
    }))
}

/// An explicit rewrite of the current HEAD commit.
#[derive(Debug, Clone)]
pub struct AmendRequest {
    /// The commit the caller reviewed. A HEAD that moved since is refused
    /// rather than rewritten into something nobody looked at.
    pub expected_head: String,
    /// Acknowledges rewriting a commit a remote-tracking ref already contains.
    pub allow_published: bool,
}

/// `git commit [--amend] -m <message> [-- <paths>]`.
///
/// With `paths` the listed files are staged first (through the same validation
/// as `stage_paths`, so nothing outside the authorized workspace can be
/// committed) and the commit is scoped to them. Without `paths` whatever is
/// already staged is committed.
///
/// `amend` rewrites history and is therefore never implied: it needs the OID
/// the caller reviewed, and a commit any remote-tracking ref already contains
/// additionally needs an explicit acknowledgement. Nothing here force-pushes.
pub fn commit(
    workspace_root: &Path,
    requested: &str,
    message: &str,
    paths: Option<&[String]>,
    amend: Option<&AmendRequest>,
) -> AppResult<CommitResult> {
    let message = message.trim();
    if message.is_empty() || message.len() > 10_000 || message.contains('\0') {
        return Err(AppError::BadRequest("Commit message is invalid".into()));
    }
    let context = require_repository(workspace_root, requested)?;
    if let Some(amend) = amend {
        let current = head_commit(workspace_root, requested)?.ok_or_else(|| {
            AppError::Conflict("There is no commit to amend on this branch".into())
        })?;
        if current.oid != amend.expected_head {
            return Err(AppError::Conflict(
                "HEAD moved since the commit was reviewed; refresh before amending".into(),
            ));
        }
        if current.published && !amend.allow_published {
            return Err(AppError::Conflict(
                "This commit is already contained in a remote-tracking ref; amending rewrites published history and needs an explicit acknowledgement".into(),
            ));
        }
    }
    let committed = match paths {
        Some(paths) => stage_paths(workspace_root, requested, paths)?.staged,
        None => vec![],
    };

    let mut args: Vec<&str> = vec!["commit"];
    if amend.is_some() {
        args.push("--amend");
    }
    args.extend(["-m", message]);
    if !committed.is_empty() {
        args.push("--");
        args.extend(committed.iter().map(String::as_str));
    }
    let mut process = command::git_command();
    process.args(&args).current_dir(&context.repository);
    let output = command::run(process, Duration::from_secs(120))?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        let stdout = String::from_utf8_lossy(&output.stdout);
        let detail = if stderr.trim().is_empty() {
            stdout.trim()
        } else {
            stderr.trim()
        };
        // "nothing to commit" is a user error, not a runtime failure.
        return Err(AppError::BadRequest(format!(
            "Git could not commit: {}",
            command::sanitize(detail)
        )));
    }
    let summary = String::from_utf8_lossy(&output.stdout).trim().to_owned();
    let commit = git(&context.repository, &["rev-parse", "--short", "HEAD"])?
        .trim()
        .to_owned();
    if let Some(amend) = amend {
        let current = git(&context.repository, &["rev-parse", "--verify", "HEAD"])?
            .trim()
            .to_owned();
        if current == amend.expected_head {
            return Err(AppError::Conflict(
                "Git reported success but HEAD still points at the original commit; inspect the repository".into(),
            ));
        }
    }
    Ok(CommitResult {
        commit,
        committed,
        summary,
    })
}

/* ---------------------------------- init ---------------------------------- */

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InitResult {
    pub repository: bool,
    /// The unborn branch `git init` selected; None if Git left HEAD detached.
    pub branch: Option<String>,
    pub path: String,
}

/// `git init` for a workspace that does not belong to any repository yet.
///
/// A directory that already resolves to a Git directory — its own, an
/// ancestor's, or a bare one — is refused instead of nested: a second
/// repository inside an existing checkout silently shadows the outer index.
/// This is the one write path with no common Git directory to queue on,
/// because that queue key only exists once the repository does.
pub fn init_repository(workspace_root: &Path) -> AppResult<InitResult> {
    let root = canonical_directory(workspace_root)?;
    let mut probe = command::git_command();
    probe
        .args(["rev-parse", "--absolute-git-dir"])
        .current_dir(&root);
    let existing = command::run(probe, Duration::from_secs(30))?;
    if existing.status.success() {
        return Err(AppError::Conflict(
            "The workspace already belongs to a Git repository".into(),
        ));
    }
    let failure = String::from_utf8_lossy(&existing.stderr);
    if !failure.contains("not a git repository") {
        return Err(AppError::Internal(command::sanitize(&failure)));
    }
    git(&root, &["init"])?;
    // Report the repository Git actually created, never the request's intent.
    let context = repo_context(&root, ".")?
        .ok_or_else(|| AppError::Internal("Git did not create a working repository here".into()))?;
    if context.repository != root {
        return Err(AppError::Internal(
            "Git initialized a repository outside the workspace root".into(),
        ));
    }
    let branch = git(
        &context.repository,
        &["symbolic-ref", "--quiet", "--short", "HEAD"],
    )
    .ok()
    .map(|value| value.trim().to_owned())
    .filter(|value| !value.is_empty());
    Ok(InitResult {
        repository: true,
        branch,
        path: workspace_relative_root(&context.repository)?,
    })
}

File: apps/desktop/src/core/git/commit.ts
import { canonicalDirectory } from "../workspaces/roots";
import { LEGACY_PREFIX, runGit } from "./command";
import { gitOutput, gitText, repoContext, requireRepository } from "./context";
import { stagePaths } from "./stage";
import { badRequest, conflict, internalError, sanitize } from "./support";

/**
 * Reading HEAD, committing (including amend) and `git init`.
 *
 * A port of `apps/runtime/src/git/commit.rs`.
 */

export interface CommitResult {
  readonly commit: string;
  readonly committed: string[];
  readonly summary: string;
}

export interface HeadCommit {
  readonly oid: string;
  readonly subject: string;
  /** The full message; empty when `truncated`. */
  readonly message: string;
  readonly truncated: boolean;
  /** A remote-tracking ref contains this commit, so rewriting it is public. */
  readonly published: boolean;
}

export interface AmendRequest {
  readonly expectedHead: string;
  readonly allowPublished: boolean;
}

export interface InitResult {
  readonly repository: boolean;
  readonly branch: string | null;
  readonly path: string;
}

/** `HEAD` as the amend controls describe it, or `null` on an unborn branch. */
export async function headCommit(
  workspaceRoot: string,
  requested: string,
): Promise<HeadCommit | null> {
  const context = await requireRepository(workspaceRoot, requested);
  const resolved = await gitOutput(context.repository, [
    "rev-parse",
    "--verify",
    "--quiet",
    "HEAD^{commit}",
  ]);
  if (resolved.status !== 0) return null;
  const oid = resolved.stdout.toString("utf8").trim();
  if (oid === "") return null;
  const raw = await gitText(context.repository, [
    "log",
    "-1",
    "--no-show-signature",
    "--format=%B",
    oid,
    "--",
  ]);
  // Trailing newlines are Git's own formatting, not part of the message.
  const message = raw.replace(/\n+$/, "");
  const truncated = Buffer.byteLength(message, "utf8") > 10_000;
  const containing = await gitText(context.repository, [
    "for-each-ref",
    "--format=%(refname)",
    "--contains",
    oid,
    "refs/remotes/",
  ]);
  return {
    oid,
    subject: message.split("\n")[0] ?? "",
    message: truncated ? "" : message,
    truncated,
    published: containing.trim() !== "",
  };
}

/**
 * `git commit [--amend] -m <message> [-- <paths>]`.
 *
 * With `paths` the listed files are staged first, through the same validation
 * as {@link stagePaths}, and the commit is scoped to them. `amend` rewrites
 * history and is therefore never implied: it needs the OID the caller
 * reviewed, and a commit any remote-tracking ref already contains
 * additionally needs an explicit acknowledgement. Nothing here force-pushes.
 */
export async function commit(
  workspaceRoot: string,
  requested: string,
  rawMessage: string,
  paths: readonly string[] | undefined,
  amend: AmendRequest | undefined,
): Promise<CommitResult> {
  const message = rawMessage.trim();
  if (
    message === "" ||
    Buffer.byteLength(message, "utf8") > 10_000 ||
    message.includes("\0")
  ) {
    throw badRequest("Commit message is invalid");
  }
  const context = await requireRepository(workspaceRoot, requested);
  if (amend !== undefined) {
    const current = await headCommit(workspaceRoot, requested);
    if (current === null) {
      throw conflict("There is no commit to amend on this branch");
    }
    if (current.oid !== amend.expectedHead) {
      throw conflict(
        "HEAD moved since the commit was reviewed; refresh before amending",
      );
    }
    if (current.published && !amend.allowPublished) {
      throw conflict(
        "This commit is already contained in a remote-tracking ref; amending rewrites published history and needs an explicit acknowledgement",
      );
    }
  }
  const committed =
    paths === undefined
      ? []
      : (await stagePaths(workspaceRoot, requested, paths)).staged;

  const args = ["commit"];
  if (amend !== undefined) args.push("--amend");
  args.push("-m", message);
  if (committed.length > 0) args.push("--", ...committed);
  const output = await runGit({
    cwd: context.repository,
    prefix: LEGACY_PREFIX,
    args,
    timeoutMs: 120_000,
  });
  if (output.status !== 0) {
    const stderr = output.stderr.toString("utf8").trim();
    const detail =
      stderr === "" ? output.stdout.toString("utf8").trim() : stderr;
    // "nothing to commit" is a user error, not a runtime failure — and git's
    // own wording for it is a screenful of porcelain (the whole untracked list
    // included), so it gets one sentence of ours instead.
    if (/nothing (added )?to commit|no changes added to commit/i.test(detail)) {
      throw badRequest("Nothing is staged to commit");
    }
    throw badRequest(`Git could not commit: ${sanitize(detail)}`);
  }
  const summary = output.stdout.toString("utf8").trim();
  const created = (
    await gitText(context.repository, ["rev-parse", "--short", "HEAD"])
  ).trim();
  if (amend !== undefined) {
    const current = (
      await gitText(context.repository, ["rev-parse", "--verify", "HEAD"])
    ).trim();
    if (current === amend.expectedHead) {
      throw conflict(
        "Git reported success but HEAD still points at the original commit; inspect the repository",
      );
    }
  }
  return { commit: created, committed, summary };
}

/**
 * `git init` for a workspace that does not belong to any repository yet.
 *
 * A directory that already resolves to a Git directory — its own, an
 * ancestor's, or a bare one — is refused instead of nested: a second
 * repository inside an existing checkout silently shadows the outer index.
 */
export async function initRepository(
  workspaceRoot: string,
): Promise<InitResult> {
  const root = canonicalDirectory(workspaceRoot);
  const existing = await gitOutput(root, ["rev-parse", "--absolute-git-dir"]);
  if (existing.status === 0) {
    throw conflict("The workspace already belongs to a Git repository");
  }
  const failure = existing.stderr.toString("utf8");
  if (!failure.includes("not a git repository")) {
    throw internalError(sanitize(failure));
  }
  await gitText(root, ["init"]);
  // Report the repository Git actually created, never the request's intent.
  const context = await repoContext(root, ".");
  if (context === undefined) {
    throw internalError("Git did not create a working repository here");
  }
  if (context.repository !== root) {
    throw internalError(
      "Git initialized a repository outside the workspace root",
    );
  }
  let branch: string | null = null;
  const symbolic = await gitOutput(context.repository, [
    "symbolic-ref",
    "--quiet",
    "--short",
    "HEAD",
  ]);
  if (symbolic.status === 0) {
    const value = symbolic.stdout.toString("utf8").trim();
    branch = value === "" ? null : value;
  }
  return { repository: true, branch, path: context.repository };
}
